packages/devextreme/js/__internal/ui/selection/m_selection.strategy.ts

```diff
@@ -33,6 +33,53 @@ export default class SelectionStrategy {
     this.options[name] = value;
   }
 
+  onSelectionChanging(): boolean | Promise<boolean> {
+    const {
+      selectedItems,
+      selectedItemKeys,
+      addedItemKeys,
+      removedItemKeys,
+      addedItems,
+      removedItems,
+      onSelectionChanging = noop,
+    } = this.options;
+
+    const selectionChangingArgs = {
+      selectedItems,
+      selectedItemKeys,
+      addedItemKeys,
+      removedItemKeys,
+      addedItems,
+      removedItems,
+      cancel: false,
+    };
+
+    onSelectionChanging(selectionChangingArgs);
+    return selectionChangingArgs.cancel;
+  }
+
+  _callCallbackIfNotCanceled(callback: () => void, cancelCallback: () => void): void {
+    const cancelResult = this.onSelectionChanging();
+
+    if (isPromise(cancelResult)) {
+      cancelResult
+        .then((cancel) => {
+          if (!cancel) {
+            callback();
+          } else {
+            cancelCallback();
+          }
+        })
+        .catch(() => {
+          callback();
+        });
+    } else if (!cancelResult) {
+      callback();
+    } else {
+      cancelCallback();
+    }
```

Comment:
minor
if you have possibility, please rewrite to async/await. Awaiting non-promise will return value as is, because of that code will be easier to read

However, maybe because of that change, datagrid qunit tests will fail :(

```suggestion
    try {
      if (await cancelResult) {
        callback();
      } else {
        cancelCallback();
      }
    } catch {
      callback();
    }
```

Reply:
It's possible but not so easy to rewrite - in your suggestion part of cases is not considered.
And also it would be strange to await this function call in parent function which works with deferred